### PR TITLE
Readd query syntax validation

### LINKED ISSUE #7411: Add best query syntax validation
See https://github.com/JabRef/jabref/pull/7391#issuecomment-768453050 for details.

### PR DESCRIPTION
**Pull request by Fu188**
Readd the query syntax validation. Fixes #7411 

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
I tried your fix locally, it does not work, it still freezes when I paste text. Also you removed the key on enter pressed.

**Comment by Fu188** [comment]
> I tried your fix locally, it does not work, it still freezes when I paste text. Also you removed the key on enter pressed.

I suppose that this issue is just to readd query syntax. In this case, I just rollback the original version and did not pull request my solution. If needed, I will pull request my solution to deal with GUI freezes.

**Comment by Fu188** [comment]
> I tried your fix locally, it does not work, it still freezes when I paste text. Also you removed the key on enter pressed.

I have upload my solution to fix GUI freezes.

File: src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java [hunk 1]
```diff
@@ -91,11 +111,67 @@ public void search() {
...
+    }
+
+    private boolean containsYearAndYearRange(String queryString) {
+        return queryString.toLowerCase().contains("year:") && queryString.toLowerCase().contains("year-range:");
+    }
+
+    // Use Thread Pool to control the match time
+    private boolean isPatternMatched(Pattern pattern, String queryString) {
+        boolean isMatched;
+        Callable<Boolean> call = () -> {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java:160]: You better use our BackGroundTask.wrap(...)

Comment (Fu188) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java]: Done! Please check it!

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java]: Yep, works better now.

**Comment by Siedlerchr** [comment]
The freezing is gone, but please readd the searching on enter. I'm just thinking if the code can be optimized a bit more. Maybe  @tobiasdiez has an idea

File: CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where groups generated from authors' last names did not include all entries of the authors' [#5833](https://github.com/JabRef/jabref/issues/5833)
 - The export to MS Office XML now uses the month name for the field `MonthAcessed` instead of the two digit number [#7354](https://github.com/JabRef/jabref/issues/7354)
 - We included some standalone dialogs from the options menu in the main preference dialog and fixed some visual issues in the preferences dialog. [#7384](https://github.com/JabRef/jabref/pull/7384)
+- Readd query syntax validation. [#7411](https://github.com/JabRef/jabref/pull/7411)
```
Comment (calixtus) [inline CHANGELOG.md:28]: I'm just thinking: If remove a feature and readd it fixed in the same release cycle, we probably should move this changelog entry to the fixed section: We fixed an issue with search freezing jabref or so...

**Comment by Fu188** [comment]
> The freezing is gone, but please readd the searching on enter. I'm just thinking if the code can be optimized a bit more. Maybe @tobiasdiez has an idea

I have readded the searching on enter. Please check it.

File: CHANGELOG.md (hunk 2 above)
Comment (Fu188) [reply CHANGELOG.md]: Done!

**Comment by koppor** [comment]
Note that we used Apache Lucene's query syntax. I am going to submit a pull request to the documentation to update it.

I think, a regular expression cannot cover Lucence's full power as lucense is a context-senstitive grammar IMHO.

I think, at the appropriate place, a try to parse the syntax **by the selected fetcher** could be tried. See here for an inspriation: https://github.com/JabRef/jabref/blob/93ad49999badd69c82b052b7d28d4df2fdf24194/src/main/java/org/jabref/logic/importer/SearchBasedFetcher.java#L42

Thus, I am strongly against adding regex syntax checking here. I would ask for parsing the query for validation with the same method the query is parsed to query the online databases.

**Review by koppor — CHANGES_REQUESTED** [review]
The code should use the "right" instance of AbstractQueryTransformer. It should enahnce these classes to provide proper feedback which part of the query is not valid.

**Comment by koppor** [comment]
Note that this PR in its current state just reverted the quick fix of https://github.com/JabRef/jabref/pull/7391. Note the following text of that PR:

grafik

I am so sorry that I did not highlight this text in #7411. I also added the screenshot there.

**Comment by koppor** [comment]
The new explanation has been created as PR to the documentation: https://github.com/JabRef/user-documentation/pull/348

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
I agree, we should reuse the query parser from lucene and catch the ParseException to construct the error message. Otherwise this looks good to me!

Thanks for reimplementing this. Next time though please ask the developers first if it is reasonable to spent time on this. I was also of the opinion that the PR that removed this feature was merged to quickly, but please respect the decision of the core developer team. We try to strongly listen to users and other contributors, and take their suggestions into account. But in the end we are the ones that have to have the bigger picture in mind and have to maintain everything in the long run. Right now you spent time reimplementing the feature, and we spent time reviewing it - and probably in two weeks @DominikVoigt will create a PR redesigning the whole search interface and making this PR obsolete. That's a bit unfortunate use of time.

Nonetheless, let's get this in and hope it will be useful for the new interface as well.

File: src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java [hunk 3]
```diff
@@ -91,11 +109,62 @@ public void search() {
...
+    private boolean containsYearAndYearRange(String queryString) {
+        return queryString.toLowerCase().contains("year:") && queryString.toLowerCase().contains("year-range:");
+    }
+
+    // Use Thread Pool to control the match time
+    private boolean isPatternMatched(Pattern pattern, String queryString) {
+        boolean isMatched = true;
+
+        try {
+            BackgroundTask.wrap(() -> pattern.matcher(queryString.strip()).matches())
```
[the remark this answers is not in the document]
Comment (Fu188) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java:160]: In this case, it's actually function `matches()` takes a lot of time that make GUI freeze. When we input a long title, it will take about 2 mintues, which will make GUI freeze

[the remark this answers is not in the document]
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java:160]: oooh that's surprising. The query parser is quicker, so this shouldn't be a problem any longer once use it.

[the remark this answers is not in the document]
Comment (Fu188) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java:160]: I am very sorry to hear this.....since I need to merge a certain number PR due to some personal reasons....
Actually, I indeed spent a lot of time dealing with this bug.

But in this milestone, I think this PR actually solve the problem of GUI freeze. If possible, I hope you can merge this request temparaily, and then consider further solution in the next milestone together!

Anyway, thanks to all developers in your team. I learned a lot from your fast responses and right guides.

[the remark this answers is not in the document]
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java:160]: If you use the query parser instead of a regex, then it's good to go from my side.

**Comment by koppor** [comment]
> Thanks for reimplementing this. Next time though please ask the developers first if it is reasonable to spent time on this. I was also of the opinion that the PR that removed this feature was merged to quickly, but please respect the decision of the core developer team. We try to strongly listen to users and other contributors, and take their suggestions into account.

It should also be noted that we never released a version without the query syntax check. We merged the PR to avoid JabRef hanging when an invalid search query was inserted

> in two weeks @DominikVoigt will create a PR redesigning the whole search interface and making this PR obsolete. 

Here, it should be noted, that the new search is already implemented. "Only" the feedback regarding the syntax is missing. 

> Nonetheless, let's get this in and hope it will be useful for the new interface as well.

The syntax parsed here is **NOT** the syntax, JabRef uses for searching. As far as I see, the regex misses AND and OR. I do not see why we should come up with a self-made regex to parse Apache Lucene.

In other words:

**The set of valid expressions detected in this PR is not the same set as the set of expression accepted by JabRef in the search**. When this PR is merged as is, the feedback to the user is wrong in many cases. (This is also the reason why we removed the validation. Besides hanging it was not completely right.)

Pleeeeeeeeeeeeeeeeeeeeease re-use the existing query parsing. I already pointed to code where it can be located from. When this is implemented, the set of validated queries is the same as JabRef takes as valid queries for searching.

For users, it would be even more useful if the position of the first error in the query syntax would be shown. This should also possible with the already included libraries.

**Comment by calixtus** [comment]
Devcall-decision: We decided to close this PR since, we haven' heard about any change here.
In fact, @DominikVoigt is already working on a different solution, using the transformators to provide feedback about parsing the query strings. So we would like to ask you to be patient for his work.
